test_plans/sycl_ext_oneapi_composite_device.asciidoc

```diff
@@ -0,0 +1,138 @@
+
+This test should take a platform the selected device belongs to and check that
+`platform::ext_oneapi_get_composite_devices()` method:
+
+* has the correct signature
+* does not throw an exception
+* returns the same result on subsequent calls
+
+==== Composite devices are not considered as root devices
+
+The test should ensure that no composite devices are returned through the
+`platform::get_devices()` method for a platform the selected device belongs to.
+
+=== New device information descriptors
+
+Composite devices are not considered as root devices and therefore they won't
+ever be selected by the CTS. To increase coverage, tests in this subcategory
+should be additionally performed on a composite device the selected device
+belongs to (if the selected device is a component device).
+
+==== `info::device::component_devices`
+
+The test should check that a query for the information descriptor has correct
+return type.
+
+The test should also check the value returned for the descriptor:
+
+* if a tested device is not a composite device, an empty vector is expected
+* otherwise, returned vector is expected to have at least two elements and those
+  elements should be additionally checked:
+** each returned device should have the `ext_oneapi_is_component` aspect
+** each returned device should return the tested device through
+   `info::device::composite_device` query
+
+==== `info::device::composite_device`
+
+The test should check that a query for the information descriptor has correct
+return type.
+
+The test should also check the value returned for the descriptor:
+
+* if a tested device is not a component device, then a synchronous
+  `exception` with `errc::invalid` error code is expected
+* otherwise, returned device should have the `ext_oneapi_is_composite` aspect
+
+=== New device aspects
+
+==== `ext_oneapi_is_composite`
+
+The test should ensure that all of devices returned by
+`get_composite_devices()` and `platform::ext_oneapi_get_composite_devices()`
+have `ext_oneapi_is_composite` aspect.
+
+==== `ext_oneapi_is_component`
+
+The test should ensure that none of devices returned by
+`get_composite_devices()` and `platform::ext_oneapi_get_composite_devices()`
+have `ext_oneapi_is_component` aspect.
+
+If the CTS selected device is a component device, then test should then attempt
+to partition a component device using one of partition methods it supports and
+check that none of sub-devices have the `ext_oneapi_is_component` aspect.
+
+=== Impact on "descendent device"
+
+The test should first obtain both a component and a composite device and it is
+skipped if that is not possible using the selected device.
+
+Having those two, the test should create a `context` using the _composite_
+device and then attempt to create a `queue` by passing the created context in
+there *and* the _component_ device. The test checks that `queue` was created
+successfully, i.e. no exceptions were thrown.
```

Comment:
It would be nice to extend this test so that it submits a kernel that uses USM.  This would be a better test of "descended device".  After the queue is created as you describe above:

* Allocate USM memory (it could be either "device" or "shared" USM) using the context that you created.
* Submit a kernel to your queue that writes into that USM memory.
* Verify that the USM memory has the expected values after the kernel completes.

I realize this expands the scope of the test, so maybe it belongs under "More complex test cases".

Reply:
I actually have it implemented [here](https://github.com/KhronosGroup/SYCL-CTS/pull/876/files#diff-d1fabf9f3122b35406b465f939761c182b9a153dcc13ffcab9e7a7b862f40d65R81) as part of "More complex test cases". I guess I can just expand that section to clarify that we specifically want to test USM as a shared resource

Reply:
It looks like that test creates a "shared context" -- a context that contains both the composite and component devices.

In order to test that "descendent devices" work properly, the context should contain only the composite device.

Reply:
Added the test case under "more complex test cases" in 7b4060b264026aaa7d3e559b4e82da8b9f5bad46